scanpipe/pipes/windows.py

```diff
@@ -0,0 +1,237 @@
+        )
+        for installed_package_file in installed_package_files:
+            installed_package_file.discovered_packages.add(created_package)
+            installed_package_file.status = "installed-package"
+            installed_package_file.save()
+        created_package.save()
+
+
+def tag_known_software(project):
+    """
+    Find Windows software in `project` by checking `project`s CodebaseResources
+    to see if their rootfs_path is is under a known software root directory. If
+    there are CodebaseResources that are under a known software root directory,
+    a DiscoveredPackage is created for that software package and all files under
+    that software package's root directory are considered installed files for
+    that package.
+
+    Currently, we are only checking for Python and openjdk in Windows Docker
+    image layers.
+    """
+    qs = project.codebaseresources.no_status()
+    python_root_directory_name_pattern = r"/Files/Python(\d*)"
+    python_root_directory_name_pattern_compiled = re.compile(
+        python_root_directory_name_pattern
+    )
+    python_paths_by_versions = {}
+    for python_codebase_resource in qs.filter(
+        rootfs_path__regex=python_root_directory_name_pattern
+    ):
+        _, version, _ = re.split(
+            python_root_directory_name_pattern_compiled,
+            python_codebase_resource.rootfs_path,
+        )
+        if not version:
+            version = "nv"
+        if version in python_paths_by_versions:
+            continue
+        if version != "nv":
+            version_with_dots = ".".join(digit for digit in version)
+            python_paths_by_versions[version_with_dots] = f"/Files/Python{version}"
+        else:
+            python_paths_by_versions[version] = "/Files/Python"
+
+    for python_version, python_path in python_paths_by_versions.items():
+        python_package = Package(
+            type="windows-program",
+            name="Python",
+            version=python_version,
+            license_expression="python",
+            copyright="Copyright (c) Python Software Foundation",
+            homepage_url="https://www.python.org/",
+        )
+        tag_installed_package_files(
+            project=project, root_dir_pattern=python_path, package=python_package
+        )
+
+    qs = project.codebaseresources.no_status()
+    openjdk_root_directory_name_pattern = r"/Files/(open)?jdk(-((\d*)(\.\d+)*))*"
+    openjdk_root_directory_name_pattern_compiled = re.compile(
+        openjdk_root_directory_name_pattern
+    )
+    openjdk_paths_by_versions = {}
+    for openjdk_codebase_resource in qs.filter(
+        rootfs_path__regex=openjdk_root_directory_name_pattern
+    ):
+        _, open_prefix, _, openjdk_version, _, _, _ = re.split(
+            openjdk_root_directory_name_pattern_compiled,
+            openjdk_codebase_resource.rootfs_path,
+        )
+        if not openjdk_version:
+            openjdk_version = "nv"
```

Comment:
We need to document with a comment... or more. This can surprising

Reply:
I've added a note to the docstring of this function